### PR TITLE
fix: consider default client when loading states

### LINKED ISSUE #308: [BUG] flaky unit test(s)
### Observed behavior

One unit test (`TestEventHandler_ProviderReadiness`) appears to be flaky, intermittently failing under certain conditions.
```
--- FAIL: TestEventHandler_ProviderReadiness (0.60s)
    --- FAIL: TestEventHandler_ProviderReadiness/for_unassociated_handler_from_default (0.20s)
        event_executor_test.go:742: timedout waiting for ready state callback, but got none
```

### Expected Behavior

All tests pass

### Steps to reproduce

Run the test e.g. 200 times:
```
go test-run ^TestEventHandler_ProviderReadiness$  -count 200
--- FAIL: TestEventHandler_ProviderReadiness (0.60s)
    --- FAIL: TestEventHandler_ProviderReadiness/for_unassociated_handler_from_default (0.20s)
        event_executor_test.go:742: timedout waiting for ready state callback, but got none
--- FAIL: TestEventHandler_ProviderReadiness (0.60s)
    --- FAIL: TestEventHandler_ProviderReadiness/for_unassociated_handler_from_default (0.20s)
        event_executor_test.go:742: timedout waiting for ready state callback, but got none
--- FAIL: TestEventHandler_ProviderReadiness (0.60s)
    --- FAIL: TestEventHandler_ProviderReadiness/for_unassociated_handler_from_default (0.20s)
        event_executor_test.go:742: timedout waiting for ready state callback, but got none
```

### PR DESCRIPTION
**Pull request by warber**
## This PR
fixes #308 and #310

### CODE REVIEW

File: openfeature/client.go [hunk 1]
```diff
@@ -713,16 +713,18 @@ func (c *Client) evaluate(
 		c.finallyHooks(ctx, hookCtx, providerInvocationClientApiHooks, options)
 	}()
 
-	// short circuit if provider is in NOT READY state
-	if c.State() == NotReadyState {
-		c.errorHooks(ctx, hookCtx, providerInvocationClientApiHooks, ProviderNotReadyError, options)
-		return evalDetails, ProviderNotReadyError
-	}
+	if _, ok := provider.(NoopProvider); !ok {
```
Comment (warber) [inline openfeature/client.go:716]: Bypass short-circuit logic for the Noop provider, as it is essentially stateless.

File: openfeature/event_executor.go [hunk 2]
```diff
@@ -213,12 +213,19 @@ func (e *eventExecutor) emitOnRegistration(domain string, providerReference prov
 	}
 }
 
-func (e *eventExecutor) State(domain string) State {
-	s, ok := e.states.Load(domain)
+func (e *eventExecutor) loadState(domain string) (State, bool) {
+	state, ok := e.states.Load(domain)
 	if !ok {
-		return NotReadyState
+		if state, ok = e.states.Load(defaultDomain); !ok {
```
Comment (warber) [inline openfeature/event_executor.go:219]: This defines the fallback logic to the default provider if no registered provider is bound

File: openfeature/openfeature_test.go [hunk 3]
```diff
@@ -718,6 +720,47 @@ func TestDefaultClientUsage(t *testing.T) {
 	}
 }
 
+func TestLateBindingOfDefaultProvider(t *testing.T) {
```
Comment (warber) [inline openfeature/openfeature_test.go:723]: I’ve added a test specifically to verify the late binding of a default provider. The first evaluation is expected to return the "default value," as a fallback to the Noop provider occurs behind the scenes. The second evaluation is expected to return the actual value from the provider that was bound through late binding.

File: openfeature/client.go (hunk 1 above)
Comment (toddbaert) [inline openfeature/client.go:716]:
```suggestion
        // bypass short-circuit logic for the Noop provider; it is essentially stateless and a "special case"
	if _, ok := provider.(NoopProvider); !ok {
```

File: openfeature/openfeature_test.go (hunk 3 above)
Comment (toddbaert) [reply openfeature/openfeature_test.go:723]: Makes sense to me, thanks for the test.

**Review by toddbaert — APPROVED** [review]
Just recommended a comment.

**Review by aepfli — APPROVED** [review]

File: openfeature/client_test.go [hunk 4]
```diff
@@ -1422,7 +1439,10 @@ func TestRequirement_1_7_7(t *testing.T) {
 		&ProviderEventing{},
 	}
 
-	_ = SetNamedProviderAndWait(t.Name(), provider)
+	err := SetNamedProviderAndWait(t.Name(), provider)
+	if err == nil {
+		t.Errorf("provider registration was expected to fail but succeeded unexpectedly")
```
Comment (beeme1mr) [inline openfeature/client_test.go:1444]: image

**Review by beeme1mr — APPROVED** [review]

**Pull request merged**